nixos/modules/services/monitoring/osquery.nix

```diff
@@ -0,0 +1,77 @@
+{ config, lib, pkgs, ... }:
+
+with builtins;
+with lib;
+
+let
+  cfg = config.services.osquery;
+  runtimeValue = flag: default:
+    cfg.config.options.${flag}
+      or cfg.flags.${flag}
+      or default;
+in
+{
+  options.services.osquery = {
+    enable = mkEnableOption (mdDoc "osqueryd daemon");
+
+    config = mkOption {
+      default = { };
+      description = mdDoc ''
+        Configuration to be written to the osqueryd JSON configuration file.
+        To understand the configuration format, refer to https://osquery.readthedocs.io/en/stable/deployment/configuration/#configuration-components.
+      '';
+      example = {
+        options.utc = false;
+      };
+      type = types.attrs;
+    };
+
+    flagfile = mkOption {
```

Comment:
I'm not certain I understand the question, sorry!
You mean as opposed to it being in the configuration file?

Maybe it's useful for me to elaborate on my thoughts for the options interface of this module. These are probably things I should also commit as comments if you agree with them. Or if you don't, we'll add document whatever we do agree on :) 

As I understand it, all configuration of osquery binaries comes through flags via the Google flags package. These flags can be either:
- in the configuration file under the `options` key
- in the flagfile specified by the special `--flagfile` command line option
- passed as individual command line options.

There is higher precedence given to flags provided in the configuration file but not all flags can be specified there. Some, like the path to the configuration file or the configuration plugin must be specified in either the flagfile or as individual command line options.

I wanted the entire configuration of osquery to be exposed to the user with as few module options as possible.
I felt that the simplest way of passing all of the flags is through the `flagfile` and the `flags` module option, simplifying our execStart to just `osquery --flagfile /path/to/flagfile`.
I've included the `flagfile` module option so that users can control this location if they choose.
Finally, the `config` option facilitate `schedule` configuration and the other aspects of the `osqueryd` configuration not controlled by flags. It can be used for setting flags if desired.

I've seen that in modules like the one for Prometheus, maintainers have painstakingly translated the Prometheus configuration file format into documented module options. I guess people find this to be a valuable addition but I would imagine it is also a lot of work to maintain that between versions. Instead, I've opted to provide the bare minimum options that expose the entire configuration to users.

I was avoiding setting default values and using application defaults so that there wasn't an additional layer of indirection to consider. Essentially so I don't have to worry if the value is coming from the application, or the package manager, or my own configuration. However, I can see that since the defaults rely on the FHS and NixOS not being that way inclined, we might to reconsider this to be more opinionated. I am on the fence about this so would happily go with whichever is more conventionally desired in NixOS modules. I'm not sure what that is though :)

Reply:
> I'm not certain I understand the question, sorry!
>
> Do you mean "Why expose this option at all?".

Yeah, my comment was a little bit too brief I guess :)

> I wanted the entire configuration of osquery to be exposed to the user with as few module options as possible.
> [...]
> I've seen that in modules like the one for Prometheus, maintainers have painstakingly translated the Prometheus configuration file format into documented module options. I guess people find this to be a valuable addition but I would imagine it is also a lot of work to maintain that between versions. Instead, I've opted to provide the bare minimum options that expose the entire configuration to users.

Agreed that we don't need to tediously copy all configuration options. I think a nice way to split the difference is something like what you're doing here: allow first-class configuration for a small set of "important" options and add an `extraConfig` for everything else.

> I wanted the entire configuration of osquery to be exposed to the user with as few module options as possible.
> I felt that the simplest way of passing all of the flags is through the flagfile and the flags module option, simplifying our execStart to just osquery --flagfile /path/to/flagfile.
> I've included the flagfile module option so that users can control this location if they choose.

But if you're configuring osquery via NixOS, you shouldn't really have a need for a `flagfile` at all, right? NixOS may (under the hood) use the `flagfile` to enact your configuration, but it's not terribly idiomatic to do "runtime configuration" in this manner. So I think your `config` is a great idea, but why allow the run-time escape hatch at all? It's a little bit at-odds with the determinism/reproducibility goals.

> However, I can see that since the defaults rely on the FHS and NixOS not being that way inclined, we might to reconsider this to be more opinionated. 

Yeah, I think respecting application defaults *unless* they're broken on Nix is probably a good compromise.

---------------------------------------------------------

Hopefully that made a bit more sense—LMK if not.

Reply:
Came across [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) which is exactly what we need I think. I added options for `database_path`, `logger_path`, and `pidfile` as a demonstration in 49d74a5.

Reply:
I resolve this discussion since @znewman01 approved the MR.